ostd/src/arch/x86/timer/apic.rs

```diff
@@ -1,76 +1,111 @@
 // SPDX-License-Identifier: MPL-2.0
 
-#![expect(unused_variables)]
-
-use alloc::sync::Arc;
 use core::{
     arch::x86_64::_rdtsc,
-    sync::atomic::{AtomicBool, AtomicU64, Ordering},
+    sync::atomic::{AtomicU64, Ordering},
 };
 
 use log::info;
-use spin::Once;
-use x86::{
-    cpuid::cpuid,
-    msr::{wrmsr, IA32_TSC_DEADLINE},
-};
 
 use super::TIMER_FREQ;
 use crate::{
     arch::{
-        kernel::tsc::init_tsc_freq,
+        kernel::apic::{self, DivideConfig},
         timer::pit::OperatingMode,
-        x86::kernel::{
-            apic::{self, DivideConfig},
-            tsc::TSC_FREQ,
-        },
+        tsc_freq,
     },
     trap::{IrqLine, TrapFrame},
 };
 
 /// Initializes APIC with tsc deadline mode or periodic mode.
-/// Return the corresponding [`IrqLine`] for the System Timer.
-pub(super) fn init() -> IrqLine {
-    init_tsc_freq();
+/// Return the corresponding [`IrqLine`] for the system timer.
+pub(super) fn init_bsp() -> IrqLine {
     if is_tsc_deadline_mode_supported() {
-        info!("[Timer]: Enable APIC TSC deadline mode.");
-        init_tsc_mode()
+        init_deadline_mode_state();
+
+        let timer_irq = IrqLine::alloc().unwrap();
+        init_tsc_mode(&timer_irq);
+        timer_irq
     } else {
-        info!("[Timer]: Enable APIC periodic mode.");
-        init_periodic_mode()
+        init_periodic_mode_state();
+
+        let timer_irq = IrqLine::alloc().unwrap();
+        init_periodic_mode(&timer_irq);
+        timer_irq
     }
 }
 
-pub(super) static APIC_TIMER_CALLBACK: Once<Arc<dyn Fn() + Sync + Send>> = Once::new();
-
 /// Determines if the current system supports tsc_deadline mode APIC timer
 fn is_tsc_deadline_mode_supported() -> bool {
+    use x86::cpuid::cpuid;
+
     const TSC_DEADLINE_MODE_SUPPORT: u32 = 1 << 24;
     let cpuid = cpuid!(1);
     (cpuid.ecx & TSC_DEADLINE_MODE_SUPPORT) > 0
 }
 
-fn init_tsc_mode() -> IrqLine {
-    let timer_irq = IrqLine::alloc().unwrap();
+pub(super) fn init_ap(timer_irq: &IrqLine) {
+    match STATE.get() {
+        Some(State::DeadlineMode { .. }) => {
+            init_tsc_mode(timer_irq);
+        }
+        Some(State::PeriodicMode { .. }) => {
+            init_periodic_mode(timer_irq);
+        }
+        None => {
+            panic!("APIC timer state not initialized");
+        }
+    }
+}
+
+fn init_tsc_mode(timer_irq: &IrqLine) {
```

Comment:
Rename this to `init_deadline_mode` for consistency. It seems that TSC mode only appears as this method name.